### ISSUE TITLE
AWS Deep Learning AMI not compatible with multi-GPU training of MolMIM

### ISSUE THREAD

**Issue by xinyu-dev** [NVIDIA/bionemo-framework#644]
## Issue

When running multi-GPU pretraining of MolMIM on AWS EC2, the following issues were observed: 

1. Pretrain on 1x A10 GPU instance on AWS works OK. 
2. Pretrain 1x L40S GPU instance + set devices:1 also works OK. 
3. Pretrain 4x A10 GPU instance + set devices:1 also works OK.
4. Pretrain on 4x A 10 GPU instance + set devices: 4 generates error similar to below.
5. Pretrain on 4x L40S GPU instance + set devices: 4  generates similar error 

Error log: 
[log.txt](https://github.com/user-attachments/files/18513694/log.txt)

## How to replicate: 

Specific AWS EC2 configurations: 
1. Launch a 4xA10 GPU instance or a 4xL40S GPU instance on AWS EC2. 
2. Choose Deep Learning Base OSS Nvidia Driver GPU AMI (Ubuntu 22.04) 20241115 (ami-0f2ad13ff5f6b6f7c) as the AMI. 
3. Run

```bash
docker run --rm -it --gpus all -p 8888:8888 nvcr.io/nvidia/clara/bionemo-framework:1.10.1 "/bin/bash"
```
4. Try running MolMIM pertaining

## Workaround

Switch to `ami-075a0f15f2d44a65e` ([NVIDIA GPU-Optimized AMI](https://aws.amazon.com/marketplace/pp/prodview-7ikjtg3um26wq?sr=0-1&ref_=beagle&applicationId=AWS-EC2-Console)). Note that some users might not have the liberty to switch AMIs therefore it might be great if we can figure out what is going on here. Thanks very much!

**Comment by jasonlin90125** [comment]
@xinyu-dev Hi! I was wondering how you obtained the pretraining files for MolMIM? I was following this tutorial: https://docs.nvidia.com/bionemo-framework/1.10/notebooks/model_training_molmim.html?highlight=molmim
I downloaded the BioNeMo Framework image but there does not appear to be a /workspace/bionemo/examples folder. Have you run into this issue before?

**Comment by xinyu-dev** [comment]
Hi, @jasonlin90125 , it has been a while since I did it. I don't remember the specifics of that folder, but here are my steps: 

1. Download and execute into this container `nvcr.io/nvidia/clara/bionemo-framework:1.10 `.  Make sure you use the tag `1.10`. 
2. Prepare the input file. Here is what my `myproject_molmim_retraining/data` folder looks like: 

<img width="214" height="177" alt="Image" src="https://github.com/user-attachments/assets/f5ca4248-1b68-40af-ac85-f69ad669fee3" />

Inside the `x000.csv`, the columns look like this: 

<img width="425" height="192" alt="Image" src="https://github.com/user-attachments/assets/d820a0dd-eab5-4fa4-9b2c-8f14168697ae" />

The `index` column doesn't need to start from 1. It can be some sort of IDs you use to retrieve the identity of the molecules in your database. It's not used during the training. 

The `canonical_smiles` column contain the canonical smiles string of each molecule you want ot use for pretraining. Converting them to canonical format is important. The original paper trained on canonical SMILES.  

If you have multiple CSV files, you must name them `x000csv`, `x001.csv`....etc

3. Below please find the config file located in `myproject_molmim_retraining/conf/pretrain_mock.yaml`

```yaml
defaults:
  - molmim_70m_24_3

name: molmim_mock_pretraining # change this to the name of the experiment 
do_training: true # do training
do_testing: false # no testing 
seed: 42
restore_from_path: null

trainer:
  devices: 4 # GPUs
  num_nodes: 1 # change to 1 node
  max_epochs: null #  the dataloader loops infinitely, so you need to set steps, not epochs
  max_steps: 340000 # check the number of steps per epoch according to the conversion notebook. You can use a larger number than calculated to manually stop the training. 
  log_every_n_steps: 100
  val_check_interval: 10000 # incrased from 2500 (too slow). validation dataset will be ran every N (this) steps. Use the epoch-step converter to estimate how many validation checks you will get within the training timeframe. E.g. 1000000 steps with 2500 steps per validation check = 400 validation checks.
  limit_train_batches: 1.0 # use all train batches. Must be a float like 1.0 not 1 (which means 1 batch)
  limit_val_batches: 1.0 # number of batches in validation step, use fraction for fraction of data, 0 to disable. Leave at 1.0 to run the entire validation dataset at the end of each epoch.
  limit_test_batches: 0 # no test dataset

exp_manager:
  name: ${name}
  exp_dir: /workspace/bionemo/myproject_molmim_retraining/results/mock # change to the path of the result folder
  explicit_log_dir: ${.exp_dir}
  create_wandb_logger: true
  create_tensorboard_logger: true
  wandb_logger_kwargs:
    project: ${name}_pretraining
    name: ${now:%Y%m%d-%H%M} # set the run name in wandb to date and time
    group: ${name}
    job_type: Localhost_nodes_${trainer.num_nodes}_gpus_${trainer.devices}
    notes: "date: ${now:%y%m%d-%H%M%S}"
    tags:
      - ${name}
    offline: False # set to True if there are issues uploading to WandB during training
  resume_if_exists: true
  resume_ignore_no_checkpoint: true
  create_checkpoint_callback: true
  checkpoint_callback_params:
    save_top_k: 3 # changed from 3 to 5
    monitor: val_molecular_accuracy
    mode: max
    save_last: true
    always_save_nemo: true # this one is not very useful. The name of the nemo file wil be re-written as ${name}.nemo which defeats the purpose. 
    filename: ${name}-${model.name}--{val_molecular_accuracy:.2f}-{val_loss:.2f}-{step}-{consumed_samples}
    model_parallel_size: ${multiply:${model.tensor_model_parallel_size}, ${model.pipeline_model_parallel_size}}
model:
  name: MolMIM-small # don't change this. There are xsmall, small, large configs
  micro_batch_size: 32 # change this
  global_batch_size: 128 # global batch size = micro batch size * data parallel size (GPUs) * accumulate_grad_batches (=1). Must not set to null
  resume_from_checkpoint: null # start trianing from random weights
  data:
    links_file: null # ${oc.env:BIONEMO_HOME}/examples/molecule/megamolbart/dataset/ZINC-downloader.txt
    dataset_path: /workspace/bionemo/myproject_molmim_retraining/data/mock # folder containing the train, test, and val files
    dataset:
      train: x000 # inclusive!
      test: x000
      val: x000
    canonicalize_target_smile: true # default training setting
    canonicalize_encoder_input: true # default training setting
    canonicalize_decoder_output: true # default training setting
    encoder_augment: false # default training setting
    decoder_independent_augment: false
    data_impl_kwargs:
      csv_mmap:
        newline_int: 10
        header_lines: 1 # header line
        workers: ${model.data.num_workers}
        sort_dataset_paths: true
        data_sep: ','
        data_col: 1 # Change this to the column index of the SMILES strings in the dataset
    use_upsampling: true
    index_mapping_dir: /workspace/bionemo/myproject_molmim_retraining/results/mock_data_index/ # index folder
  optim:
    name: fused_adam
    lr: 0.0005
    weight_decay: 0.001
    betas:
    - 0.9
    - 0.999
    sched:
      name: CosineAnnealing
      warmup_steps: ${multiply:${trainer.max_steps}, 0.01}
      constant_steps: ${multiply:${trainer.max_steps}, 0.05}
      max_steps: ${trainer.max_steps}
      min_lr: 5.0e-05
```

This is the same setting as what is used to train the MolMIM weights, except that the GPU resources. 

4. Create a bash file wit this content: 
```bash
#! /bin/bash

set -e

# CD into direcotry
cd /workspace/bionemo/myproject_molmim_retraining/pretrain

# remove index folder
rm -rf /workspace/bionemo/myproject_molmim_retraining/results/mock_data_index/

# remove if exists
rm -f dlloger.json

# remove results folder
rm -rf /workspace/bionemo/myproject_molmim_retraining/results/mock/



# train
python pretrain.py \
--config-path /workspace/bionemo/myproject_molmim_retraining/conf \
--config-name pretrain_mock.yaml
```

Update the path accordingly depend your file structure. Then run the bash file to start training. It will ask for you wandb key during the first run but remembers it for the later runs.

**Comment by jasonlin90125** [comment]
@xinyu-dev Thank you so much! This is very helpful.

**Comment by trvachov** [comment]
Closing as stale — this concerned gradient accumulation in MolMIM. The `bionemo-molmim` sub-package has been completely removed from the repository.

**Comment by trvachov** [comment]
Hi @xinyu-dev — thank you for this detailed report, and for the helpful walkthrough you provided to @jasonlin90125!

MolMIM was part of the legacy BioNeMo Framework 1.x (the NGC container `nvcr.io/nvidia/clara/bionemo-framework:1.10.1`), which is a separate product from this open-source repository (BioNeMo Framework 2.x). MolMIM has been fully removed from this codebase and is not available in the current framework.

The multi-GPU issue you encountered appears to be specific to the AWS Deep Learning Base AMI NCCL/driver configuration rather than a BioNeMo code bug — as evidenced by the fix being an AMI swap to the NVIDIA GPU-Optimized AMI.

Since this issue pertains to the legacy 1.x container (no longer maintained here) and an AWS AMI environment configuration, we are closing this as out of scope for the current repository. For questions about the legacy BioNeMo 1.x container, please reach out via [NVIDIA Developer Forums](https://forums.developer.nvidia.com/).

Thank you for your contribution!
